docker-compose.yaml

```diff
@@ -0,0 +1,155 @@
+# Copyright 2018 Cargill Incorporated
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+version: '3.6'
+
+services:
+
+  settings-tp:
+    build:
+      context: .
+      dockerfile: families/settings/Dockerfile
+    image: sawtooth-settings-tp-local:${ISOLATION_ID}
+    volumes:
+      - ./:/project/sawtooth-core
+    container_name: sawtooth-settings-tp-local
+    depends_on:
+      - validator
+    command: |
+      bash -c "
+        bin/protogen
+        cd families/settings
+        python3 setup.py clean --all
+        python3 setup.py build
+        settings-tp -vv -C tcp://validator:4004
+      "
+    stop_signal: SIGKILL
+
+
+  intkey-tp-python:
+    build:
+      context: .
+      dockerfile: sdk/examples/intkey_python/Dockerfile
+    image: intkey-tp-python-local:${ISOLATION_ID}
+    volumes:
+      - ./:/project/sawtooth-core
+    container_name: intkey-tp-python-local
+    depends_on:
+      - validator
+    command: |
+      bash -c "
+        bin/protogen
+        cd sdk/examples/intkey_python
+        python3 setup.py clean --all
+        python3 setup.py build
+        intkey-tp-python -vv -C tcp://validator:4004
+      "
+    stop_signal: SIGKILL
+
+
+  xo-tp-python:
+    build:
+      context: .
+      dockerfile: sdk/examples/xo_python/Dockerfile
+    image: xo-tp-python-local:${ISOLATION_ID}
+    volumes:
+      - ./:/project/sawtooth-core
+    container_name: xo-tp-python-local
+    depends_on:
+      - validator
+    command: |
+      bash -c "
+        bin/protogen
+        cd sdk/examples/xo_python
+        python3 setup.py clean --all
+        python3 setup.py build
+        xo-tp-python -vv -C tcp://validator:4004
+      "
+    stop_signal: SIGKILL
+
+
+  client:
+    build:
+      context: .
+      dockerfile: docker/sawtooth-shell
+    image: sawtooth-shell-local:${ISOLATION_ID}
+    volumes:
+      - ./:/project/sawtooth-core
+    container_name: sawtooth-shell-local
+    depends_on:
+      - validator
+    command: |
+      bash -c "
+        sawtooth keygen && \
+        tail -f /dev/null
+      "
+    stop_signal: SIGKILL
+
+
+  validator:
+    build:
+      context: .
+      dockerfile: validator/Dockerfile
+    image: sawtooth-validator-local:${ISOLATION_ID}
+    volumes:
+      - ./:/project/sawtooth-core
```

Comment:
How do we feel about masking the `/target` directory so we don't dump a bunch of build artifacts onto our host machines?

As seen here:
https://github.com/hyperledger/sawtooth-supply-chain/blob/master/docker-compose.yaml#L70

Reply:
While I think that is a cool way to avoid cluttering up your directory locally, I think we should keep the default behavior as simple and intuitive as possible. This is the "mounted" docker-compose file so I would expect it to just mount the directory, not mount the directory except directories A, B, C, etc. Now we are asking people to dig through the compose file to see what directories really are mounted and which are masked.

It also seems like starting down this road would open the door for extending this list every time someone decided they did not want a particular directory modified locally. Seems better to have a simple default and let people customize the behavior based their preferences.

Reply:
I think "don't mount generated files" is a pretty unambiguous rule, and does not open the door to any extensions (unless they are more generated files). 

More important than spamming repos with unusable generated files, if people are doing local work you will overwrite their generated files, causing difficult to diagnose bugs. The reverse is true as well. Running `cargo build` on your host machine will break a currently running container. I think preventing these bugs is more important than the confusion that may result if people intentionally try to use the target directory as a mounted directory. I don't understand what the use case for that would be, so why would it ever come up?

Reply:
I think that what are describing is a matter of opinion. My understanding of the goal of this update to the build system is to make the way it works as simple and intuitive as possible. The goal of the "mounted" Dockerfile is to provide an environment that you can mount the source directory into and run it. The simplest and most intuitive thing that can be done to achieve this goal is to mount the source repo as-is. If there are users who have more complex needs then I think they should solve them as needed, rather than try to accommodate every need out of the box. The "accommodate everyone" philosophy is what got the build tools into the nasty state they are in now in the first place.

Reply:
In other words, there is nothing stopping you from adding that line yourself when you are running things locally, if that is the behavior you want. I just think the out of the box behavior should be as simple and intuitive as possible.